Prometheus metric exporter

hectorhdzg opened the pull request:
This implementation is based on [OpenCensus Prometheus exporter](https://github.com/census-instrumentation/opencensus-python/tree/master/contrib/opencensus-ext-prometheus)

I added a lot of code from #341 to unblock testing and lint for this, once Metrics code is updated this PR only need to merge with latest changes. Please review Prometheus related changes and add comments in PR 341 PR for Metric related stuff.

There are still TODOs regarding supporting histograms and adding tests for Gauges and Measures, this is currently blocked by missing Aggregators in opentelemetry SDK

Fixes #157

ocelotl commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,150 @@
+_NON_LETTERS_NOR_DIGITS_RE = re.compile(r"[^\w]", re.UNICODE | re.IGNORECASE)
+
+
+def sanitize(key):
```
This is called in only one place, maybe put this code directly there instead of leaving something importable at module level?

hectorhdzg replied:
sanitize need to be applied to label names as well, added that missing logic

ocelotl commented on opentelemetry-api/src/opentelemetry/metrics/__init__.py:
```diff
@@ -171,48 +151,42 @@ def get_handle(self, label_set: LabelSet) -> "MeasureHandle":
 
 
 # pylint: disable=unused-argument
-class Meter(abc.ABC):
```
Why remove this?

mauriciovasquezbernal commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,150 @@
+        for example when the HTTP endpoint is invoked by Prometheus.
+        """
+
+        for metric_batch in list(self._metrics_to_export):
```
Out of curiosity, what is the purpose of copying the list before iterating?

hectorhdzg replied:
Not needed removed

mauriciovasquezbernal commented on ext/opentelemetry-ext-prometheus/README.rst:
```diff
@@ -0,0 +1,71 @@
+
+.. code:: python
+
+    from opentelemetry import metrics
```
It'd be nice to have this example as a Python file, so one can execute it without copy-paste.

mauriciovasquezbernal commented on ext/opentelemetry-ext-prometheus/README.rst:
```diff
@@ -0,0 +1,71 @@
+
+    counter.add(25, label_set)
+    # We sleep for 5 seconds, exported value should be 25
+    time.sleep(5)
```
Btw, I'm not sure this sleep makes sense after https://github.com/open-telemetry/opentelemetry-python/pull/341#discussion_r376161304, it could be better to wait for user to press enter to finish or something similar?

hectorhdzg commented:
@mauriciovasquezbernal @ocelotl I tried to address all comments please let me know if you have more feedback, we can update examples to use new PullController when available

mauriciovasquezbernal commented on ext/opentelemetry-ext-prometheus/tests/__init__.py:
```diff
@@ -0,0 +1,13 @@
+# Copyright 2020, OpenTelemetry Authors
```
We could just let this file empty.

mauriciovasquezbernal commented on ext/opentelemetry-ext-prometheus/tests/test_prometheus_exporter.py:
```diff
@@ -0,0 +1,119 @@
+
+    def test_gauge_to_prometheus(self):
+        # TODO: Add unit test once GaugeAggregator is available
+        self.assertEqual(True, True)
```
`pass`?

ocelotl replied:
Even better, remove the test function and just leave the `TODO` comment. Pytest still collects, runs and reports this test as it was meaningful when it is not.

mauriciovasquezbernal commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/export/batcher.py:
```diff
@@ -97,4 +97,4 @@ def process(self, record):
             aggregator = self.aggregator_for(record.metric.__class__)
             aggregator.merge(record.aggregator)
-        self._batch_map[batch_key] = aggregator
+        self._batch_map[batch_key] = aggregator
```
Remove this change from this PR.

mauriciovasquezbernal commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/export/controller.py:
```diff
@@ -53,4 +53,4 @@ def tick(self):
         self.exporter.export(self.meter.batcher.checkpoint_set())
         # Perform post-exporting logic based on batcher configuration
-        self.meter.batcher.finished_collection()
+        self.meter.batcher.finished_collection()
```
Same.

mauriciovasquezbernal commented on ext/opentelemetry-ext-prometheus/tests/test_prometheus_exporter.py:
```diff
@@ -0,0 +1,119 @@
+
+    def test_measure_to_prometheus(self):
+        # TODO: Add unit test once Measure Aggregators are available
+        self.assertEqual(True, True)
```
`pass`?

ocelotl replied:
Same here.

mauriciovasquezbernal commented on ext/opentelemetry-ext-prometheus/tests/test_prometheus_exporter.py:
```diff
@@ -0,0 +1,119 @@
+    def test_counter_to_prometheus(self):
+        meter = metrics.Meter()
+        metric = meter.create_metric(
+            "test@name",
```
I'd prefer to have also an explicitly test for `sanitize()`.

mauriciovasquezbernal commented on ext/opentelemetry-ext-prometheus/tests/test_prometheus_exporter.py:
```diff
@@ -0,0 +1,119 @@
+            self.assertEqual(
+                prometheus_metric.samples[0].labels["environment_"], "staging"
+            )
+
```
Is it missing to test that a warning is printed when an unsupported metrics type is received?

mauriciovasquezbernal commented on ext/opentelemetry-ext-prometheus/tests/test_prometheus_exporter.py:
```diff
@@ -0,0 +1,119 @@
+            self.assertEqual(type(prometheus_metric), CounterMetricFamily)
+            self.assertEqual(prometheus_metric.name, "testprefix_test_name")
+            self.assertEqual(prometheus_metric.documentation, "testdesc")
+            self.assertEqual(prometheus_metric.samples[0].value, 123)
```
Check that `len(prometheus_metric.samples[0]) == 1`?

mauriciovasquezbernal commented on ext/opentelemetry-ext-prometheus/tests/test_prometheus_exporter.py:
```diff
@@ -0,0 +1,119 @@
+            self.assertEqual(prometheus_metric.documentation, "testdesc")
+            self.assertEqual(prometheus_metric.samples[0].value, 123)
+            self.assertEqual(
+                prometheus_metric.samples[0].labels["environment_"], "staging"
```
Also here, check len of labels?

mauriciovasquezbernal commented on docs/opentelemetry.metrics.rst:
```diff
@@ -4,4 +4,4 @@ opentelemetry.metrics package
 ---------------
 
-.. automodule:: opentelemetry.metrics
+.. automodule:: opentelemetry.metrics
```
Remove this change.

mauriciovasquezbernal commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,157 @@
+                labels=label_values, value=metric_record.aggregator.check_point
+            )
+
+        # TODO: Add support for histograms when supported in OT
```
Should we print a warning just in case?

hectorhdzg replied:
Hard to check for specific condition for the warning because the Measure type doesn't exist yet

ocelotl commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,150 @@
+    """
+
+    def __init__(
+        self, port: int = DEFAULT_PORT, address: str = "", prefix: str = ""
```
Sorry, this is marked as resolved, but I don't see a change here, do you still want to change it?

ocelotl commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,157 @@
+                )
+                if prometheus_metric:
+                    yield prometheus_metric
+            self._metrics_to_export.remove(metric_batch)
```
This specific piece of code probably works fine, but modifying a sequence while iterating over it can lead to very weird bugs.

ocelotl replied:
I suggest using a `while` here:
```python
while self._metrics_to_export:
    for metric_record in self._metrics_to_export.popleft():
        ...
    ...
```

ocelotl commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,157 @@
+                prometheus_metric = self._translate_to_prometheus(
+                    metric_record
+                )
+                if prometheus_metric:
```
Use `is not None` here.

ocelotl commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,157 @@
+
+    def __init__(self, prefix: str = ""):
+        self._prefix = prefix
+        self._metrics_to_export = []
```
Don't you mean `add_metrics_data` could use `append` while collect could use `popleft`?

mauriciovasquezbernal replied:
What's the difference between using `appendleft` & `pop` vs `append` & `popleft`?

ocelotl replied:
I mean, is just a matter of keeping what the current implementation is doing right now. `add_metrics_data` appends to the right of `self._metrics_to_export` and collect is "popping" (actually, it is calling `remove`) from left to right.

hectorhdzg replied:
Added append and popleft let me know if there are strong opinions about this

ocelotl commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,157 @@
+        label_values = []
+        label_keys = []
+        for label in metric_record.label_set.labels:
+            for index, label_tuple_value in enumerate(label):
```
Instead of calling `enumerate` and then checking on `index` you can just use a slice and step to get only the odd elements of `label`:

```python
for element in ["a", "b", "c", "d", "e"][1::2]:
    print(element)
```
```
b
d
```

hectorhdzg commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,157 @@
+            metric_name = self._prefix + "_"
+        metric_name += sanitize(metric_record.metric.name)
+
+        if isinstance(metric_record.metric, Counter):
```
This looks pretty cool, but I'm not sure I understand how this works exactly can you explain this a little bit more?

With this code I'm currently getting a KeyError: <class 'opentelemetry.sdk.metrics.Counter'>

ocelotl replied:
Ok, a solution is:

```python
        if isinstance(metric_record.metric, (Counter, Gauge, Measure)):
            prometheus_metric = {
                Counter: CounterMetricFamily,
                Gauge: GaugeMetricFamily,
                Measure: UnknownMetricFamily
            }[metric_record.metric.__class__.__bases__[1]](
                name=metric_name,
                documentation=metric_record.metric.description,
                labels=label_keys,
            )
            prometheus_metric.add_metric(
                labels=label_values, value=metric_record.aggregator.check_point
            )

        else:
            logger.warning(
                "Unsupported metric type. %s", type(metric_record.metric)
            )
```

This was failing because the class of `metric_record.metric` was a child of `Gauge` instead of being `Gauge` itself. It is just a suggestion, you may not find it worthy to dig that deep into the attributes of an object to avoid some repeated code.

hectorhdzg replied:
Let me know if you have strong opinions about this one, I agree we will have less repeated code but we are adding more complicated one.

ocelotl replied:
Not really, feel free to use this suggestion or not as you like :wink:

mauriciovasquezbernal commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,154 @@
+                labels=label_keys,
+            )
+            prometheus_metric.add_metric(
+                labels=label_values, value=metric_record.aggregator.check_point
```
```suggestion
                labels=label_values, value=metric_record.aggregator.checkpoint
```

mauriciovasquezbernal commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,154 @@
+                labels=label_keys,
+            )
+            prometheus_metric.add_metric(
+                labels=label_values, value=metric_record.aggregator.check_point
```
```suggestion
                labels=label_values, value=metric_record.aggregator.checkpoint
```

mauriciovasquezbernal commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,154 @@
+                labels=label_keys,
+            )
+            prometheus_metric.add_metric(
+                labels=label_values, value=metric_record.aggregator.check_point
```
```suggestion
                labels=label_values, value=metric_record.aggregator.checkpoint
```

hectorhdzg commented:
@mauriciovasquezbernal are you getting any error when running the example?, did you run pip install -e .\opentelemetry-ext-prometheus\?

mauriciovasquezbernal commented:
@hectorhdzg I'm sorry for not being clear. It was not working for me because the `check_point` vs `checkpoint`, now that you have renamed it, it should be fine.

lzchen commented on ext/opentelemetry-ext-prometheus/README.rst:
```diff
@@ -0,0 +1,71 @@
+
+    counter.add(25, label_set)
+    # We sleep for 5 seconds, exported value should be 25
+    time.sleep(5)
```
Still missing `time`?

lzchen commented on ext/opentelemetry-ext-prometheus/CHANGELOG.md:
```diff
@@ -0,0 +1,4 @@
+# Changelog
+
+## Unreleased
+
```
Are we leaving this blank?

hectorhdzg replied:
Is not blank it says unreleased, once is released changelogs will make sense

toumorokoshi replied:
@c24t are we doing changelogs on this fashion? I wonder if there's an easier way to manage these than attempting to update changelogs in all repos as we go along. Then again, this may play better with pypi.

c24t replied:
This is how we started the other changelogs. The idea is that each PR that makes a relevant change should add a line under "unreleased", and those lines get moved under a release-specific header when we cut the release.

lzchen commented on ext/opentelemetry-ext-prometheus/setup.cfg:
```diff
@@ -0,0 +1,47 @@
+install_requires =
+    prometheus_client >= 0.5.0, < 1.0.0
+    opentelemetry-api
+    opentelemetry-sdk
```
Do we want to have a min version of the sdk that includes the metrics sdk changes?

hectorhdzg replied:
It doesn't exist yet, we can add it when is available

lzchen commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/version.py:
```diff
@@ -0,0 +1,15 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__version__ = "0.4.dev0"
```
Are we synchronizing the versions to the api and sdk packages?

hectorhdzg replied:
Yes we are doing this in all other ext packages

lzchen commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/version.py:
```diff
@@ -0,0 +1,15 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__version__ = "0.4.dev0"
```
Add license info to `tests/__init__.py`.

hectorhdzg replied:
I believe it was @mauriciovasquezbernal who asked me to remove it, file is empty so not sure if license makes sense

mauriciovasquezbernal replied:
@hectorhdzg I asked to remove it from an empty __init__.py file: https://github.com/open-telemetry/opentelemetry-python/pull/378#discussion_r378236599.
I think this one should include the license.

lzchen commented on examples/metrics/prometheus.py:
```diff
@@ -0,0 +1,51 @@
+
+# Meter is responsible for creating and recording metrics
+meter = Meter()
+metrics.set_preferred_meter_implementation(meter)
```
I believe `set_preferred_meter_implementation` takes in a `meter_factory` rather than an actual `meter`.

https://github.com/open-telemetry/opentelemetry-python/blob/master/examples/metrics/record.py#L27

lzchen commented on examples/metrics/prometheus.py:
```diff
@@ -0,0 +1,51 @@
+counter = meter.create_metric(
+    "incoming requests",
+    "incoming requests",
+    "bytes",
```
Is this the correct unit for incoming requests?

Also, see [this comment](https://github.com/open-telemetry/opentelemetry-python/pull/341#discussion_r376758482)

lzchen commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,154 @@
+    Args:
+        port: Port number to listen.
+        address: Endpoint address (default is localhost).
+        prefix: single-word application prefix relevant to the domain the metric belongs to.
```
Nit: Line wrapping

lzchen commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,154 @@
+    def collect(self):
+        """Collect fetches the metrics from OpenTelemetry
+        and delivers them as Prometheus Metrics.
+        Collect is invoked every time a prometheus.Gatherer is run
```
Is there a link to some API that shows how this works?

hectorhdzg replied:
The custom collector will use same logic as regular Prometheus Collector, pulling mechanism is configured by time intervals but is also triggered by other input like the user refreshing Prometheus visualization UI

https://github.com/prometheus/client_python#custom-collectors
https://github.com/prometheus/client_python/blob/master/prometheus_client/registry.py

lzchen commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,154 @@
+        label_values = []
+        label_keys = []
+        for label in metric_record.label_set.labels:
+            for label_tuple_value in label[1::2]:
```
I believe you can do `label[1]` when accessing one element of a tuple.

lzchen commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,154 @@
+                labels=label_keys,
+            )
+            prometheus_metric.add_metric(
+                labels=label_values, value=metric_record.aggregator.checkpoint
```
How does Prometheus handle discrepancies with missing label_keys/label values? Do they expect the same amount of keys and values passed in, and if so, we probably have to do some preprocessing here.

e.g1. label_keys: ["key1", "key2"], label_values: ["value1"]
eg2. label_keys: ["key1"], label_values: ["value1", "value2"]

hectorhdzg replied:
Prometheus will handle this correctly, if two labels are provided and only one value it will only add the populated one, same thing the other way around, I will update this logic to use MetricRecord.label_set.labels to be sure every single label key have a value

lzchen replied:
Nice! Also seems like we might remove the recommended [keys](https://github.com/open-telemetry/opentelemetry-specification/issues/463) since Prometheus does not use this in their protocol.

lzchen commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,154 @@
+                label_values.append(label_tuple_value)
+
+        for label_key in metric_record.metric.label_keys:
+            label_keys.append(self._sanitize(label_key))
```
Are we using the label keys supplied from the label set or the recommended keys in a metric instrument as shown here?
According to [the specs](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-metrics.md#withrecommendedkeys-declaration-on-metric-instruments), recommended keys should be used if its relevant for the exporter.

hectorhdzg replied:
Updated to use label set key instead of recommended keys

lzchen commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,154 @@
+        for label_key in metric_record.metric.label_keys:
+            label_keys.append(self._sanitize(label_key))
+        metric_name = ""
+        if self._prefix != "":
```
You can do `if self._prefix:` I believe

lzchen commented:
Looks good! Just some questions about how the label keys and label values behavior is handled.

mauriciovasquezbernal commented on examples/metrics/prometheus.py:
```diff
@@ -0,0 +1,51 @@
+# Copyright 2020, OpenTelemetry Authors
```
I think this example could be integrated in the already existing metrics examples likes we do in tracing: https://github.com/open-telemetry/opentelemetry-python/blob/96943b242fc56b56e65695fc092bcbee7b32b27f/examples/basic_tracer/tracer.py#L27-L36

Having multiple / very similar examples is a maintenance nightmare.

hectorhdzg replied:
Created #423 to track this

hectorhdzg commented:
@ocelotl let me know if you have more feedback for this one

toumorokoshi commented on ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py:
```diff
@@ -0,0 +1,153 @@
+        self._address = address
+        self._collector = CustomCollector(prefix)
+
+        start_http_server(port=self._port, addr=self._address)
```
what is the value of being opinionated around the way that prometheus metrics are served?

It seems like the prometheus_client itself provides multiple ways to integrate prometheus endpoints in to an application, some of which are more desirable (e.g. integrating the route directly into a wsgi / asgi app vs starting a whole separate http server):

https://github.com/prometheus/client_python#wsgi

hectorhdzg replied:
I agree we can add a flag to start the server on by default allowing customers to use any Prometheus client. Let me know what you think

hectorhdzg replied:
Removed start_server call entirely instead of using parameter

hectorhdzg commented:
I removed start_http server call from exporter, users can use any kind of Prometheus client and we just attach our Collector to the Prometheus registry, added extra call in README and sample so people can see their data when testing the exporter. @toumorokoshi, @codeboten , @lzchen  and @ocelotl  please let me know if this ok for everyone.
